### PR TITLE
[Python] Non-deterministic format-spec matching

### LINKED ISSUE #2232: [Python] Incorrect format context in string
- Sublime Version: 4061
- OS Version: Ubuntu 19.10

Given the following Python dictionary:
```python
{
    'diagnostics':
    [
        {
            'message': "Argument of type '{ props: { translations: { type: (ObjectConstructor | FunctionConstructor)[]; required: boolean; }; }; data: () => { showBanner: boolean; fallback: { title: string; text: string; agreeButtonText: string; closeTitle: string; }; }; computed: { ...; }; mounted: typeof mounted; methods: { ...; }; }' is not assignable to parameter of type 'new (...args: any[]) => any'.\n  Type '{ props: { translations: { type: (ObjectConstructor | FunctionConstructor)[]; required: boolean; }; }; data: () => { showBanner: boolean; fallback: { title: string; text: string; agreeButtonText: string; closeTitle: string; }; }; computed: { ...; }; mounted: typeof mounted; methods: { ...; }; }' provides no match for the signature 'new (...args: any[]): any'.",
            'code': 2345,
            'severity': 1,
            'range':
            {
                'end':
                {
                    'character': 0,
                    'line': 0
                },
                'start':
                {
                    'character': 0,
                    'line': 0
                }
            },
            'source': 'Vetur'
        }
    ],
    'uri': 'file:///Users/cookie.vue'
}
```

The value for the `message` key starts a `constant.other.placeholder.python` scope whereas it should not.

I believe a minimal example for this would be

```python
x="{:{ {}}"
```

My Python interpreter says this is not a formatting string:

```
>>> x="{:{ {}}".format("1", "2", "3")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: unmatched '{' in format spec
```

### LINKED ISSUE #2241: [Python] Breaking on strings with { and !a inside single quotes
The source code is from: 
* https://github.com/python/cpython/blob/91874bb07161bb481b6f5ea18ffafe69cb8cac30/Lib/test/test_logging.py#L3864-L3878

```Python
        self.assertRaises(ValueError, logging.Formatter, '{{foo!r:4.2}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{{foo!r:4.2}}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo/bar}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo:{{w}}.{{p}}}}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo!X:{{w}}.{{p}}}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo!a:random}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo!a:ran{dom}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo!a:ran{d}om}', style='{')
        self.assertRaises(ValueError, logging.Formatter, '{foo.!a:d}', style='{')

        # Dollar style
        # Testing failure for mismatch bare $
        self.assert_error_message(
            ValueError,
            "invalid format: bare \'$\' not allowed",
            logging.Formatter, '$bar $$$', style='$'
        )
        self.assert_error_message(
            ValueError,**
```

image

After that point, the scope becomes all `source.python meta.function-call.arguments.python meta.string.python string.quoted.single.python constant.other.placeholder.python meta.format-spec.python constant.other.format-spec.python`

### Environment

* Operating system and version:
  * Windows `10 build 15063 x64`
  * Resolution `1920x1080`
  * dpi_scale `1.0`
  * Sublime Text:
  * Build `3207`
  * 64 bit

### PR DESCRIPTION
**Pull request by FichteFoll**
Correctly match some tricky format specifications using branches. Some
quite unusual ones are not correctly matched but tests have been added
to document the exact behavior.

My first experiement with non-determinism and quite a tricky case at that because we're not aiming for 100% accuracy but it still allows for more correct highlighting.

Closes #2232.
Closes #2241.

### CODE REVIEW

File: Python/syntax_test_python_strings.py [hunk 1]
```diff
@@ -549,6 +549,45 @@
...
+# ^ - constant.other.placeholder
+
+# Syntactically correct, but hardly come up in real code
+"{:{ ()}}".format(0, **{" ()": "d"}) == '0'
+# ^ constant.other.placeholder
+'{foo/bar}'.format(**{"foo/bar": 1}) == '1'
+# ^ - constant.other.placeholder
+
+# Legal but non-standard format spec
+'{foo:{{w}}.{{p}}}}'  # illegal double {{ in format spec
```
Comment (wbond) [inline Python/syntax_test_python_strings.py:584]: This line confuses me. Above it says "Legal but non-standard", whereas this line says "illegal".

File: Python/syntax_test_python_strings.py [hunk 2]
```diff
@@ -549,6 +549,45 @@
...
+# ^ - constant.other.placeholder
+'{foo!a:random}'
+# ^ - constant.other.placeholder
```
Comment (wbond) [inline Python/syntax_test_python_strings.py:587]: Is this another case of a legal format that the current implementation doesn't handle?

File: Python/Python.sublime-syntax [hunk 3]
```diff
@@ -1410,22 +1412,38 @@ contexts:
...
         - match: '![ars]'
           scope: storage.modifier.conversion.python
         - match: ':'
           push:
             - meta_scope: meta.format-spec.python constant.other.format-spec.python
             - match: (?=\})
               pop: true
-            - include: formatting-syntax
+            - match: (?=\{)
+              push: formatting-syntax-complex  # need to push extra because of pop: 2
```
Comment (wbond) [inline Python/Python.sublime-syntax:1444]: Could this be worked around using `set` on line 1429?

File: Python/syntax_test_python_strings.py (hunk 1 above)
Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:584]: The syntax is legal and translates to `format(foo, "{w}.{p}")`. The comment refers to how a format spec of `{w}.{p}` is unusual and must be interpreted by a custom `__format__` implementation, because this is unheard of in the entire stdlib. I'll reword it to be clearer.

Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:584]: Actually, the spec is illegal because `}` (or rather `}}` in the escaped form) doesn't work. This doesn't seem to be documented anywhere, but the following snippet can be used to test behavior:
```py
class Foo:
    def __format__(self, spec):
        return spec.format(w=1, p=123)
fmt = '{foo:{{w}}.{{p}}}}'
print(fmt.format(foo=Foo()))
```
```
Traceback (most recent call last):
  File "/tmp/format-spec.py", line 5, in <module>
    print(fmt.format(foo=Foo()))
ValueError: Single '}' encountered in format string
```

Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:584]: Here's an example implementation showing how it works:
```py
class Foo:
    def __format__(self, spec):
        return spec.format(w=1, p=123)
fmt = '{foo:{{w}}.{{p}}}'
print(fmt.format(foo=Foo()))
```
```
1.123
```

File: Python/syntax_test_python_strings.py (hunk 2 above)
Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:587]: Yes, but deliberately.

File: Python/syntax_test_python_strings.py [hunk 4]
```diff
@@ -549,6 +549,45 @@
...
+'{foo.!a:d}'
+# ^ constant.other.placeholder
+"{:{ {}}"  # issue 2232
+# ^ - constant.other.placeholder
+
+# Syntactically correct, but hardly come up in real code
+"{:{ ()}}".format(0, **{" ()": "d"}) == '0'
+# ^ constant.other.placeholder
+'{foo/bar}'.format(**{"foo/bar": 1}) == '1'
+# ^ - constant.other.placeholder
```
[the remark this answers is not in the document]
Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:581]: The opening comment explains why we don't need or want to match all legal syntax.
```
# We don't actually want to match the syntax 100% of the time,
# since we never know for sure if the string is used as a format string,
# so some of these matches are incorrect because of implementation details.
```

[the remark this answers is not in the document]
Comment (wbond) [reply Python/syntax_test_python_strings.py:581]: Right, it just isn't crystal clear to me which is which, based on some of the ambiguous comments. I just figured it would be good to know on a case by case basis which is a non-match due to the implementation.

File: Python/syntax_test_python_strings.py [hunk 5]
```diff
@@ -549,6 +549,45 @@
...
+'{foo!a:ran{dom}'  # unclosed
+# ^ - constant.other.placeholder
+'{foo!a:ran{dom}'  # unclosed
+# ^ - constant.other.placeholder
+
+# Invalid field names
+'{foo{d}}'
+# ^ - constant.other.placeholder
+'{foo.!a:d}'
+# ^ constant.other.placeholder
```
[the remark this answers is not in the document]
Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:573]: That is correct-ish. Trying to format this raises `ValueError: Empty attribute in format string`, but in reality it's the same as typing `foo.` outside of a format string which we still highlight because of a) convenience for the highlighter and b) highlighting while typing.

File: Python/Python.sublime-syntax (hunk 3 above)
Comment (FichteFoll) [reply Python/Python.sublime-syntax:1444]: Wasn't possible before because I couldn't `set` out of the `\{` match due to a recently fixed issue in the syntax, but I'll try now.

File: Python/syntax_test_python_strings.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FichteFoll) [reply Python/syntax_test_python_strings.py:581]: I have a commit updating the comments locally. Will push once I got to the `set` part.

**Comment by FichteFoll** [comment]
I refined the commits a but and removed the pop:2. The extra push is still there, but this time because it *should* be there and should generally behave better in one particular edge case.

**Comment by wbond** [comment]
Thanks for the extra info!
